ltdconveyor/s3upload.py

```diff
@@ -0,0 +1,395 @@
+
+    For immutable uploads simply using ``cache_control`` is more efficient
+    since it allows the browser to also locally cache content.
+
+    .. seelso:
+
+       - `Fastly: Cache control tutorial
+         <https://docs.fastly.com/guides/tutorials/cache-control-tutorial>`_.
+       - `Google: HTTP caching <http://ls.st/39v>`_.
+    """
+    log.debug('s3upload.upload({0}, {1}, {2})'.format(
+        bucket_name, path_prefix, source_dir))
+
+    session = boto3.session.Session(
+        profile_name=aws_profile,
+        aws_access_key_id=aws_access_key_id,
+        aws_secret_access_key=aws_secret_access_key)
+    s3 = session.resource('s3')
+    bucket = s3.Bucket(bucket_name)
+
+    metadata = None
+    if surrogate_key is not None:
+        if metadata is None:
+            metadata = {}
+        metadata['surrogate-key'] = surrogate_key
+    if surrogate_control is not None:
+        if metadata is None:
+            metadata = {}
+        metadata['surrogate-control'] = surrogate_control
+
+    manager = ObjectManager(session, bucket_name, path_prefix)
+
+    for (rootdir, dirnames, filenames) in os.walk(source_dir):
+        # name of root directory on S3 bucket
+        bucket_root = os.path.relpath(rootdir, start=source_dir)
+        if bucket_root in ('.', '/'):
+            bucket_root = ''
+
+        # Delete bucket directories that no longer exist in source
+        bucket_dirnames = manager.list_dirnames_in_directory(bucket_root)
+        for bucket_dirname in bucket_dirnames:
+            if bucket_dirname not in dirnames:
+                log.debug(('Deleting bucket directory {0}'.format(
+                    bucket_dirname)))
+                manager.delete_directory(bucket_dirname)
+
+        # Delete files that no longer exist in source
+        bucket_filenames = manager.list_filenames_in_directory(bucket_root)
+        for bucket_filename in bucket_filenames:
+            if bucket_filename not in filenames:
+                bucket_filename = os.path.join(bucket_root, bucket_filename)
+                log.debug('Deleting bucket file {0}'.format(bucket_filename))
+                manager.delete_file(bucket_filename)
+
+        # Upload files in directory
+        for filename in filenames:
+            local_path = os.path.join(rootdir, filename)
+            bucket_path = os.path.join(path_prefix, bucket_root, filename)
+            log.debug('Uploading to {0}'.format(bucket_path))
+            upload_file(local_path, bucket_path, bucket,
+                        metadata=metadata, acl=acl,
+                        cache_control=cache_control)
+
+        # Upload a directory redirect object
+        if upload_dir_redirect_objects is True:
+            bucket_dir_path = os.path.join(path_prefix, bucket_root)
+            bucket_dir_path = bucket_dir_path.rstrip('/')
+            if metadata:
+                redirect_metadata = dict(metadata)
```

Comment:
Here I was specifically trying to make a copy of `metadata` since `redirect_metadata` needs to be modified with additional keys. But in general I'm looking back at `metadata` initialization and see that it can be tidied up. 👍 Also `if metadata` -> `if metadata is not None` is PEP8.